public_html/wp-content/plugins/pattern-directory/tests/phpunit/favorite-test.php

```diff
@@ -0,0 +1,160 @@
+<?php
+/**
+ * Test Block Pattern validation.
+ */
+
+use function WordPressdotorg\Pattern_Directory\Favorite\{save_favorite, get_favorites};
+use const WordPressdotorg\Pattern_Directory\Favorite\META_KEY;
+use const WordPressdotorg\Pattern_Directory\Pattern_Post_Type\POST_TYPE;
+
+/**
+ * Test pattern validation.
+ */
+class Pattern_Favorite_Test extends WP_UnitTestCase {
+	protected static $pattern_id;
+	protected static $faved_pattern_id;
+	protected static $page_id;
+	protected static $user_admin;
+	protected static $user_subscriber;
+
+	/**
+	 * Setup fixtures that are shared across all tests.
+	 */
+	public static function wpSetUpBeforeClass( $factory ) {
+		self::$pattern_id = $factory->post->create(
+			array( 'post_type' => POST_TYPE )
+		);
+		self::$faved_pattern_id = $factory->post->create(
+			array( 'post_type' => POST_TYPE )
+		);
+		self::$page_id = $factory->post->create(
+			array( 'post_type' => 'page' )
+		);
+		self::$user_admin = $factory->user->create(
+			array(
+				'role' => 'administrator',
+			)
+		);
+		self::$user_subscriber = $factory->user->create(
+			array(
+				'role' => 'subscriber',
+			)
+		);
+		update_user_meta( self::$user_admin, META_KEY, array( self::$faved_pattern_id ) );
+	}
+
+	/**
+	 * Test favoriting a pattern as the admin user.
+	 */
+	public function test_favorite_pattern() {
+		wp_set_current_user( self::$user_admin );
+		$success = save_favorite( self::$pattern_id );
+		$this->assertTrue( (bool) $success );
+
+		$favorites = get_user_meta( self::$user_admin, META_KEY, true );
+		$this->assertSame( $favorites, array( self::$faved_pattern_id, self::$pattern_id ) );
+	}
+
+	/**
+	 * Test unfavoriting a pattern as the admin user.
+	 */
+	public function test_unfavorite_faved_pattern() {
+		wp_set_current_user( self::$user_admin );
+		$success = save_favorite( self::$faved_pattern_id, null, false );
```

Comment:
Yeah, I agree it's not super intuitive - I admit I grabbed the base for this from plugin favoriting. I thought since we'll only be interacting with this via the API (which does have separate add/remove functionality), it would be abstracted enough.

Reply:
It would be a bit more intuitive if the favorites were saved as separate user meta values instead of as a single serialized array. Then you would use `add_user_meta` and `delete_user_meta` instead of `update`, and you wouldn't necessarily need the boolean. But I'm not sure if there's a compelling reason to switch to that now. It doesn't seem like there would be a big performance difference either way, unless someone was favoriting thousands of patterns.

Reply:
Actually, thinking about this in light of us wanting to be able to show how many times a specific pattern has been favorited, it will probably be easier and more efficient to switch to storing these in individual user meta rows, instead of as a serialized array...

Reply:
Switched to separate user meta rows, using add_user_meta/delete_user_meta — agreed that this will be easier to query for per-pattern in the future. I also split the save_ function into `add_favorite` and `remove_favorite`, since there really is no reason to keep it confusing 🙂